addons/account/models/account_account.py

```diff
@@ -811,18 +794,114 @@ def _get_most_frequent_account_for_partner(self, company_id, partner_id, move_ty
+        extra_vals = {
+            'recent_accounts': suggested_accounts,
+        }
+
+        def _records_to_list(records, highlight=None):
+            if template:
+                return [
+                    (
+                        record.id,
+                        record.display_name,
+                        record._to_qweb(template, highlight, extra_values=extra_vals),
+                    )
+                    for record in records
+                ]
             return [(record.id, record.display_name) for record in records]
-        return super().name_search(name, domain, operator, limit)
+
+        if not name and suggested_accounts:
+            records = self.sudo().browse(suggested_accounts)
+            return _records_to_list(records)
+
+        search_domain = Domain('display_name', 'ilike', name) if name else []
+
+        move_type_accounts = defaultdict(list)
+        move_type_accounts.update({
+            'out': ['income'],
+            'in': ['expense', 'asset'],
+        })
+        move_type_prefix = move_type.split('_')[0]
+        # search all account types if the search contains a number
+        digit_in_search_term = any(c.isdigit() for c in name)
+        internal_group_domain = [('internal_group', 'in', move_type_accounts[move_type_prefix])] if not digit_in_search_term else []
+
+        domain = expression.AND([search_domain, internal_group_domain, domain])
+        records = self.with_context(preferred_account_ids=suggested_accounts).search(domain, limit=limit)
+        return _records_to_list(records, highlight=name)
+
+    @api.model
+    @api.readonly
+    def web_name_search(self, name, specification, domain=None, operator='ilike', limit=100):
+        if self.env.context.get('option_template_ref'):
+            record_tuples = self.name_search(name, domain, operator, limit)
+            return [
+                {
+                    'id': id_,
+                    'display_name': display_name,
+                    'html': html,
+                }
+                for id_, display_name, html in record_tuples
+            ]
+        return super().web_name_search(name, specification, domain, operator, limit)
+
+    def _to_qweb(self, template=None, highlight=None, tag='span', extra_values={}):
+        self.ensure_one()
+        re_search_term = self._get_search_term_regex(highlight)
+        if not template:
+            content = self._highlight(self.display_name, re_search_term) if highlight else self.display_name
+            return Markup(f"<{escape(tag)}>{content}</{escape(tag)}>")
+
+        qweb_content = self.env['ir.qweb']._render(
+            template=template,
+            values={
+                'record': self,
+                **extra_values,
+            },
+        )
+        return self._highlight(qweb_content, re_search_term) if highlight else qweb_content
+
+    @functools.lru_cache
```

Comment:
Does the caching really change something?

Reply:
Apparently `compile` takes up most of the time, so I thought caching might be a good idea